[CUMULUS-2812] Remove DynamoDB logic from rules PUT

botanical opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-2812:  Remove DynamoDB logic from rules PUT](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2812)

## Changes
- Remove DynamoDB logic from rules `PUT` endpoint
- Move event resources logic from `rulesModel` to `rulesHelper`
- Add `translateApiRuleToPostgresRuleRaw`

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

botanical commented on packages/db/src/translate/rules.ts:
```diff
@@ -54,36 +54,56 @@ export const translatePostgresRuleToApiRule = async (
+    dbClient,
+    collectionPgModel,
+    providerPgModel
+  );
```
These were copied from this PR https://github.com/nasa/cumulus/pull/2733/files

markdboyd replied:
We should copy the unit tests too

botanical replied:
[[`43f847a` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/43f847a0c56a3e95f7fc8d7b7d3c01acc8538c0c)](url)

markdboyd commented on packages/api/endpoints/rules.js:
```diff
@@ -137,48 +141,38 @@ async function put(req, res) {
-    const oldRule = await ruleModel.get({ name });
+    const oldRule = await rulePgModel.get(knex, { name });
+    const oldApiRule = await translatePostgresRuleToApiRule(oldRule, knex);
+    const translatedApiRule = await translateApiRuleToPostgresRuleRaw(apiRule, knex);
```
```suggestion
    const apiPgRule = await translateApiRuleToPostgresRuleRaw(apiRule, knex);
```

This is a nit, but changing this variable name helps me follow the logic a bit more

botanical replied:
[`b891209` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/b891209fba38568d9aa132371d7cb4d886730ee6)

markdboyd commented on packages/api/endpoints/rules.js:
```diff
@@ -137,48 +141,38 @@ async function put(req, res) {
   try {
-    const oldRule = await ruleModel.get({ name });
+    const oldRule = await rulePgModel.get(knex, { name });
+    const oldApiRule = await translatePostgresRuleToApiRule(oldRule, knex);
```
Let's move this inside the if block on line 160 since it it's only necessary in the case of `rerun`

botanical replied:
[`b891209` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/b891209fba38568d9aa132371d7cb4d886730ee6)

markdboyd commented on packages/api/tests/endpoints/test-rules.js:
```diff
@@ -729,14 +734,16 @@ test('PUT replaces a rule', async (t) => {
     createdAt: Date.now(),
     updatedAt: Date.now(),
   };
+  console.log('UPDATE RULE', updateRule);
```
Remove this

botanical replied:
[`4c12492` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/4c1249251939ede5e8b8d6fe90b051ca72effc6d)

markdboyd commented on packages/api/tests/endpoints/test-rules.js:
```diff
@@ -745,39 +752,19 @@ test('PUT replaces a rule', async (t) => {
   t.is(actualPostgresRule.created_at.getTime(), updatedEsRecord.createdAt);
   t.is(actualPostgresRule.updated_at.getTime(), updatedEsRecord.updatedAt);
-
-  t.like(actualPostgresRule, {
-    ...omit(originalPgRecord, ['queue_url']),
-    enabled: true,
-    created_at: new Date(originalDynamoRule.createdAt),
-    updated_at: actualPostgresRule.updated_at,
-  });
```
I think we want to keep this test to verify that the originalPgRule.created_at was preserved

botanical replied:
[`4c12492` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/4c1249251939ede5e8b8d6fe90b051ca72effc6d#diff-49139a123843cc0118f2bc2b3275724314181f1de87be662e46c5125aca6c844R771)

markdboyd commented on packages/api/tests/endpoints/test-rules.js:
```diff
@@ -745,39 +752,19 @@ test('PUT replaces a rule', async (t) => {
+  const actualPostgresRule = await rulePgModel.get(testKnex, { name: updateRule.name });
+  const updatedEsRecord = await esRulesClient.get(translatedPgRecord.name);
 
-  t.true(actualRule.updatedAt > updateRule.updatedAt);
```
I think we want to keep this as:

`t.true(actualPostgresRule.updated_at > originalPgRecord.updated_at);`

botanical replied:
[`4c12492` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/4c1249251939ede5e8b8d6fe90b051ca72effc6d#diff-49139a123843cc0118f2bc2b3275724314181f1de87be662e46c5125aca6c844R758)

markdboyd commented on packages/api/tests/lib/rules/test-createRuleResources.js:
```diff
@@ -0,0 +1,358 @@
+  const consumerEventMappings = kinesisEventMappings[0].EventSourceMappings;
+  const logEventMappings = kinesisEventMappings[1].EventSourceMappings;
+
+  console.log(JSON.stringify(kinesisEventMappings));
```
Remove this

botanical replied:
[`907d615` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/907d6158686dd2f2b5bc640e15c153114fce40ae)

botanical replied:
[`6e4273e` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/6e4273e419e493d91939e1f3fc8ea21e59aa6d22)

markdboyd commented on packages/api/tests/lib/rules/test-createRuleResources.js:
```diff
@@ -0,0 +1,358 @@
+    },
+  });
+  const sqsRule = await createRuleTrigger(rule);
+  t.deepEqual(sqsRule, rule);
```
I don't know if this is proving anything? Since you're including the `meta` in the `rule` fixture, there's no way to know whether that property was added by the `validateAndUpdateSqsRule` function. I would remove `meta` from the `rule` fixture then compare `sqsRule` against a rule with the expected `meta`

markdboyd replied:
I see that the test below proves that the expected `meta` gets set. So is this test necessary?

botanical replied:
Hmm the first test uses values for meta fields that aren't default. But I can see your perspective

markdboyd replied:
Fair enough. I can see that now. Thanks for clarifying

markdboyd commented on packages/api/tests/lib/rules/test-createRuleResources.js:
```diff
@@ -0,0 +1,358 @@
+    type: 'sqs',
+    value: queues.queueUrl,
+    enabled: true,
+    meta: {},
```
This shouldn't be necessary, I believe

botanical replied:
[`907d615` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/907d6158686dd2f2b5bc640e15c153114fce40ae#diff-f5c5b96c6230da09644104acbdca2268e22d32509bc5b275a5b4648a8939584eL236)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+
+  t.is(createdRule.name, kinesisRule.name);
+  t.is(createdRule.value, kinesisRule.value);
+  t.false(createdRule.arn === undefined);
+  t.false(createdRule.log_event_arn === undefined);
```
```suggestion
  t.not(createdRule.arn, undefined);
  t.not(createdRule.log_event_arn, undefined);
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+  });
+  // create rule trigger
+  const createdRule = await createRuleTrigger(kinesisRule);
+  t.false(createdRule.arn === undefined);
+  t.false(createdRule.log_event_arn === undefined);
```
```suggestion
  t.not(createdRule.arn, undefined);
  t.not(createdRule.log_event_arn, undefined);
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+    updateRuleTrigger(sqsRuleWithTrigger, updatedKinesisRule, t.context.testKnex),
+    { name: 'SchemaValidationError' }
+  );
+  const updatedKinesisEventMappings = await getKinesisEventMappings();
+  const updatedConsumerEventMappings = updatedKinesisEventMappings[0].EventSourceMappings;
+  const updatedLogEventMappings = updatedKinesisEventMappings[1].EventSourceMappings;
+
+  t.is(updatedConsumerEventMappings.length, 0);
+  t.is(updatedLogEventMappings.length, 0);
```
None of this is relevant for SQS rules

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40L299) 🤦

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+  });
+});
+
+test.serial('Updating a valid SQS rule to have an invalid schema throws an error and does not update triggers', async (t) => {
```
```suggestion
test.serial('Updating a valid SQS rule to have an invalid schema throws an error', async (t) => {
```

There are no triggers for SQS rules that get updated on the rule (no `arn` property is set)

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40L299)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+  });
+});
+
+test.serial('Updating a rule trigger for an SQS rule fails if there is no redrive policy on the queue', async (t) => {
```
```suggestion
test.serial('Updating an SQS rule fails if there is no redrive policy on the queue', async (t) => {
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40R306)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+  t.teardown(async () => await SQS.deleteQueue(queueUrl));
+});
+
+test.serial('Updating a rule trigger for an SQS rule succeeds', async (t) => {
```
```suggestion
test.serial('Updating queue for an SQS rule succeeds', async (t) => {
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40R331)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+    value: newQueues.queueUrl,
+  };
+  const updatedSqsRule = await updateRuleTrigger(sqsRule, updatedRule, t.context.testKnex);
+  t.deepEqual(updatedSqsRule.value, newQueues.queueUrl);
```
```suggestion
  t.is(updatedSqsRule.value, newQueues.queueUrl);
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40R352)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+        promise: () => Promise.resolve({
+          Subscriptions: [{
+            Endpoint: process.env.messageConsumer,
+            SubscriptionArn: snsTopicArn,
```
```suggestion
            SubscriptionArn: randomString(),
```

The subscription ARN for the Lambda subscription to the SNS topic is separate from the SNS topic ARN

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40R436)

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+
+  const ruleWithTrigger = await createRuleTrigger(rule);
+
+  t.is(rule.value, TopicArn);
```
```suggestion
  t.truthy(rule.arn);
```

markdboyd commented on packages/api/tests/lib/rules/test-updateRuleResources.js:
```diff
@@ -0,0 +1,475 @@
+
+  const ruleWithTrigger = await createRuleTrigger(rule);
+
+  t.is(rule.value, TopicArn);
```
```suggestion
  t.is(rule.value, TopicArn);
  t.truthy(rule.arn);
```
```

botanical replied:
[`aa59c90` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/aa59c90a42ee176d4075b4de2c1bf78678177b09#diff-057c67806291c57c6410d819b57bda8d87b5370152a299e6baca44b5a44f0e40R403-R404)

markdboyd reviewed (changes requested):
Overall this looks good, but a few requested changes and comments

botanical commented on packages/api/endpoints/rules.js:
```diff
@@ -137,48 +141,38 @@ async function put(req, res) {
+    const translatedApiRule = await translateApiRuleToPostgresRuleRaw(apiRule, knex);
 
     apiRule.updatedAt = Date.now();
     apiRule.createdAt = oldRule.createdAt;
```
🙇

botanical replied:
[`b891209` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/b891209fba38568d9aa132371d7cb4d886730ee6)

botanical commented on packages/api/lib/rulesHelpers.js:
```diff
@@ -261,7 +268,407 @@ async function deleteRuleResources(knex, rule) {
+    error.detail = 'Rule workflow is undefined.';
+    throw error;
+  }
+  if (!rule.type) {
+    error.detail = 'Rule type is undefined.';
+    throw error;
+  }
```
`createRuleTrigger` sets the default value for `enabled` to be true if it is undefined. Did you still want me to add that extra condition? It will almost certainly always be set. 🤔

markdboyd replied:
Did we add these unit tests for the endpoint?

botanical commented on packages/api/lib/rulesHelpers.js:
```diff
@@ -261,7 +268,407 @@ async function deleteRuleResources(knex, rule) {
+async function createRuleTrigger(ruleItem) {
+  let newRuleItem = cloneDeep(ruleItem);
+  // the default value for enabled is true
+  if (ruleItem.enabled === undefined) {
```
[`faedb8b` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/faedb8b4151419009c1850b7580eeebb75e86e8a)

botanical commented on packages/api/tests/lib/rules/test-createRuleResources.js:
```diff
@@ -0,0 +1,358 @@
+  t.false(createdRule.log_event_arn === undefined);
+
+  // Clean Up
+  // await deleteRuleResources(t.context.testKnex, kinesisRule);
```
[`907d615` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/907d6158686dd2f2b5bc640e15c153114fce40ae)

botanical commented on packages/db/src/models/rule.ts:
```diff
@@ -19,7 +19,7 @@ class RulePgModel extends BasePgModel<PostgresRule, PostgresRuleRecord> {
       .onConflict('name')
       .merge()
-      .returning('cumulus_id');
+      .returning('*');
```
[`5adf919` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/5adf919e8bafd101518a1dbfe02bbd725ef15ba6)

botanical commented on example/spec/parallel/testAPI/snsRuleSpec.js:
```diff
@@ -267,13 +267,20 @@ describe('The SNS-type rule', () => {
     beforeAll(async () => {
       if (beforeAllFailed) return;
       try {
+        // need to explicitly set arn to null because including arn isn't allowed when setting a disabled rule to enabled
+        const updateParams = {
+          ...updatedRule,
+          state: 'ENABLED',
```
I was confused as well and am not entirely sure how it worked before to be honest.

botanical commented on example/spec/parallel/testAPI/snsRuleSpec.js:
```diff
@@ -267,13 +267,19 @@ describe('The SNS-type rule', () => {
     beforeAll(async () => {
       if (beforeAllFailed) return;
       try {
+        // need to explicitly set arn to null because including arn isn't allowed when setting a disabled rule to enabled
```
remove this

botanical replied:
```suggestion
```

markdboyd commented on packages/api/lib/rulesHelpers.js:
```diff
@@ -531,21 +531,21 @@ async function buildPayload(item) {
-    queueUrl: item.queueUrl,
-    asyncOperationId: item.asyncOperationId,
-    executionNamePrefix: item.executionNamePrefix,
+    provider: rule.provider,
+    collection: rule.collection,
```
These properties don't exist on a Postgres rule, as far as I know? There would only be `rule.provider_cumulus_id` and `rule.collection_cumulus_id`

botanical replied:
That's a really good point. I've updated the buildPayload function to translate the PG rule, since the translate functions makes a call to look up the provider and collection [`bb74e3a` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/bb74e3aa1d4642235116c170a15beaa6bb63eb93#diff-803bdbf08060a60307d82cb3938c90378a3c359858b4decb1d9089444c92d57bR523)

botanical replied:
Sorry, there were a couple changes I made after that last commit https://github.com/nasa/cumulus/pull/2871/files#diff-803bdbf08060a60307d82cb3938c90378a3c359858b4decb1d9089444c92d57bR519 The gist is that I updated buildPayload to take in an ApiRule and all calls to `buildPayload` need to translate a PG rule to an API rule. Also, I made `asyncOperationId` and `cumulus_meta` fields optional parameters to the function.

markdboyd replied:
I proposed a refactor to the logic here:

https://github.com/nasa/cumulus/pull/2871/files?show-viewed-files=true&file-filters%5B%5D=#r821139656

markdboyd commented on packages/api/tests/lib/rules/test-createRuleResources.js:
```diff
@@ -352,3 +355,50 @@ test.serial('Creating an enabled SNS rule creates an event source mapping', asyn
+  };
+  const rule = fakeRuleRecordFactory(ruleParams);
+  const expectedPayload = {
+    provider: rule.provider,
+    collection: rule.collection,
```
I think this test is passing because `rule.provider` and `rule.collection` are undefined on the `payload` as well. So the test passes but maybe doesn't prove what we want?

botanical replied:
True [`bb74e3a` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/bb74e3aa1d4642235116c170a15beaa6bb63eb93#diff-f5c5b96c6230da09644104acbdca2268e22d32509bc5b275a5b4648a8939584eR363)

markdboyd commented on example/spec/parallel/testAPI/snsRuleSpec.js:
```diff
@@ -267,13 +267,18 @@ describe('The SNS-type rule', () => {
+        const updateParams = {
+          ...updatedRule,
+          state: 'ENABLED',
+          rule: {
+            value: updatedRule.rule.value,
+            type: 'sns',
+          },
```
Since we don't have to set `rule.arn` explicitly, we probably don't need to update the `rule` property at all?

markdboyd commented on packages/api/endpoints/rules.js:
```diff
@@ -137,48 +141,38 @@ async function put(req, res) {
     // If rule type is onetime no change is allowed unless it is a rerun
     if (apiRule.action === 'rerun') {
-      return models.Rule.invoke(oldRule).then(() => res.send(oldRule));
+      return invokeRerun(oldApiRule).then(() => res.send(oldApiRule));
     }
 
-    const fieldsToDelete = Object.keys(oldRule).filter(
-      (key) => !(key in apiRule) && key !== 'createdAt'
-    );
-    const postgresRule = await translateApiRuleToPostgresRule(apiRule, knex);
+    const ruleWithTrigger = await updateRuleTrigger(oldRule, apiPgRule, knex);
```
What about this refactor:

```
    // If rule type is onetime no change is allowed unless it is a rerun
    if (apiRule.action === 'rerun') {
      return invokeRerun(oldApiRule).then(() => res.send(oldApiRule));
    }

    const apiRuleWithTrigger = await updateRuleTrigger(oldApiRule, apiRule, knex);
    const apiPgRule = await translateApiRuleToPostgresRuleRaw(apiRuleWithTrigger, knex);
```

The key change is moving the call to `translateApiRuleToPostgresRuleRaw`up here. The real benefit of these changes would be downstream in `createRuleTrigger` and `updateRuleTrigger`, which would now handle an API style rule instead of a PG rule. It seems like this refactor would eliminate the need to use `translatePostgresRuleToApiRule` before calling `buildPayload`? 

In essence, the only translation would live in the endpoint, since are primarily writing to PG but returning "API" style records. I think it would reduce the overall number of translation calls/DB queries. And perhaps keep the API helper code simpler to always handle an API rule?

botanical replied:
I think that's reasonable. I updated `createRuleTrigger` and `updateRuleTrigger` to handle an API style rule and also updated the put endpoint 

[`68155b3` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/68155b3331587a648cfa8a71f3b863c9bfd173fb)

botanical commented on packages/api/lib/rulesHelpers.js:
```diff
@@ -608,8 +609,10 @@ async function updateRuleTrigger(ruleItem, updates, knex) {
+  case 'onetime':
     break;
+  default:
+    throw new ValidationError(`Rule type \'${mergedRule.rule.type}\' not supported.`);
```
I made these changes and added `onetime` as a case because of this unit test I wrote to check that PUT errors when a rule type is invalid [`68155b3` (#2871)](https://github.com/nasa/cumulus/pull/2871/commits/68155b3331587a648cfa8a71f3b863c9bfd173fb#diff-49139a123843cc0118f2bc2b3275724314181f1de87be662e46c5125aca6c844R973) 

I can add `scheduled` as a rule type if we want to keep this functionality. Otherwise, I can remove the unit test if we shouldn't be checking for valid rule types in `updateRuleTrigger`

botanical commented:
I added the following tests to `test-rules`
```
PUT returns a 400 response if record is missing workflow property
PUT returns a 400 response if record is missing type property
PUT returns a 400 response if rule name is invalid
PUT returns a 400 response if rule type is invalid
```
and updated the POST unit tests to check the error message/details

markdboyd commented on packages/api/lib/rulesHelpers.js:
```diff
@@ -273,25 +274,25 @@ async function deleteRuleResources(knex, rule) {
 function updateSnsRuleArn(ruleItem, snsSubscriptionArn) {
   const updatedRuleItem = cloneDeep(ruleItem);
-  updatedRuleItem.arn = snsSubscriptionArn;
+  updatedRuleItem.rule.arn = snsSubscriptionArn;
```
I was wondering if we should revert this back to the previous version of the logic since we're handling an API rule now, but if the unit tests are passing maybe we don't need to worry about it

markdboyd replied:
Nope we shouldn't

markdboyd reviewed (approved):
Great work! Thanks for your patience with my feedback!
